enterprise/server/remote_execution/casfs/casfs.go

```diff
@@ -0,0 +1,653 @@
+			child := &Node{cfs: cfs, parent: parentNode, symlinkTarget: childSymlinkNode.GetTarget()}
+			inode := cfs.root.NewPersistentInode(ctx, child, fs.StableAttr{Mode: fuse.S_IFLNK})
+			if !parentNode.AddChild(childSymlinkNode.Name, inode, false) {
+				return status.UnknownErrorf("could not add child %q", childSymlinkNode.Name)
+			}
+			if cfs.verbose {
+				log.Debugf("[%s] Input symlink: %s", cfs.taskID(), child.relativePath())
+			}
+		}
+		return nil
+	}
+
+	err = walkDir(rootDirectory, cfs.root)
+	if err != nil {
+		return err
+	}
+
+	outputDirs := make(map[string]struct{})
+	for _, dir := range fsLayout.OutputDirs {
+		if cfs.verbose {
+			log.Debugf("[%s] Output dir: %s", cfs.taskID(), dir)
+		}
+		outputDirs[dir] = struct{}{}
+	}
+	for _, file := range fsLayout.OutputFiles {
+		if cfs.verbose {
+			log.Debugf("[%s] Output file: %s", cfs.taskID(), file)
+		}
+		outputDirs[filepath.Dir(file)] = struct{}{}
+	}
+	for dir := range outputDirs {
+		parts := strings.Split(dir, string(os.PathSeparator))
+		node := cfs.root
+		for _, p := range parts {
+			childNode := node.GetChild(p)
+			if childNode == nil {
+				child := &Node{cfs: cfs, parent: node}
+				inode := cfs.root.NewPersistentInode(ctx, child, fs.StableAttr{Mode: fuse.S_IFDIR})
+				if !node.AddChild(p, inode, false) {
+					return status.UnknownErrorf("could not add child %q", p)
+				}
+				node = child
+			} else {
+				node = childNode.Operations().(*Node)
+			}
+		}
+	}
+
+	cfs.fileFetcher = fileFetcher
+
+	return nil
+}
+
+func (cfs *CASFS) fetchFiles(filesToFetch dirtools.FileMap, opts *dirtools.DownloadTreeOpts) error {
+	cfs.mu.Lock()
+	ff := cfs.fileFetcher
+	taskID := cfs.internalTaskID
+	cfs.mu.Unlock()
+	if ff == nil {
+		return status.FailedPreconditionErrorf("[%s] file fetcher is not set", taskID)
+	}
+	return ff.FetchFiles(filesToFetch, opts)
+}
+
+func (cfs *CASFS) taskID() string {
+	cfs.mu.Lock()
+	defer cfs.mu.Unlock()
+	return cfs.internalTaskID
+}
+
+func (cfs *CASFS) CountReadBytes(n int) {
+	cfs.mu.Lock()
+	defer cfs.mu.Unlock()
+	cfs.bytesRead += n
+}
+
+func (cfs *CASFS) CountWrittenBytes(n int) {
+	cfs.mu.Lock()
+	defer cfs.mu.Unlock()
+	cfs.bytesWritten += n
+}
+
+func (cfs *CASFS) FinishTask() {
+	cfs.mu.Lock()
+	defer cfs.mu.Unlock()
+	cfs.fileFetcher = nil
+	cfs.internalTaskID = "unset"
+}
+
+func (cfs *CASFS) Unmount() error {
+	if cfs.server == nil {
+		return nil
+	}
+
+	cfs.mu.Lock()
+	readBytes := cfs.bytesRead
+	wroteBytes := cfs.bytesWritten
+	cfs.mu.Unlock()
+
+	log.Infof("[%s] Unmounting. Total read %s, wrote %s.", cfs.taskID(), units.HumanSize(float64(readBytes)), units.HumanSize(float64(wroteBytes)))
+
+	return cfs.server.Unmount()
+}
+
+type Node struct {
+	fs.Inode
+
+	cfs    *CASFS
+	parent *Node
+
+	// Whether this node represents an input to the action.
+	// Modifications to inputs are denied.
+	isInput bool
+
+	fileNode      *repb.FileNode
+	symlinkTarget string
+
+	mu          sync.Mutex
+	scratchFile string
```

Comment:
nit: consider calling this `scratchFilePath` (and maybe rename `scratchPath` below to `computeScratchPath` if it seems ambiguous)